Production/src/BaseTupleProducer.cc

```diff
@@ -323,20 +377,49 @@ void BaseTupleProducer::FillGenParticleInfo()
 {
     using analysis::GenEventType;
     static constexpr int electronPdgId = 11, muonPdgId = 13, tauPdgId = 15;
+    static const std::set<int> leptons = { 11, 13, 15 };
     static const std::set<int> bosons = { 23, 24, 25, 35 };
 
     std::vector<const reco::GenParticle*> particles_to_store;
+    std::map<int,std::vector<const reco::GenParticle*>> particles_to_store_backup;
 
     std::map<int, size_t> particle_counts;
     for(const auto& particle : *genParticles) {
         const auto& flag = particle.statusFlags();
-        if(!flag.isPrompt() || !flag.isLastCopy()) continue;
+        if(!flag.isPrompt() || !flag.isLastCopy() || !flag.fromHardProcess()) continue;
         const int abs_pdg = std::abs(particle.pdgId());
         ++particle_counts[abs_pdg];
-        if(saveGenBosonInfo && bosons.count(abs_pdg))
-            particles_to_store.push_back(&particle);
+        if(saveGenBosonInfo && bosons.count(abs_pdg)){
+          particles_to_store.push_back(&particle);
+        }
+        if(leptons.count(abs_pdg)){
+          particles_to_store_backup[abs_pdg].push_back(&particle);
+        }
+
+    }
+
+    if(particles_to_store.empty()){
```

Comment:
this logic should be only applied only when saveGenBosonInfo is true, or I'm mistaken?